src/redisraft.c

```diff
@@ -1677,8 +1743,15 @@ void handleClientEvent(RedisModuleCtx *ctx, RedisModuleEvent eid,
                  * We can only append if we are leader, if we're not leader anymore,
                  * then this would have already been cleaned up.
                  */
-                if (cs && cs->watched && rr->raft && raft_is_leader(rr->raft)) {
-                    appendEndClientSession(rr, NULL, ci->id, "disconnect");
+                if (cs) {
+                    if (cs->watched && rr->raft && raft_is_leader(rr->raft)) {
+                        appendEndClientSession(rr, NULL, ci->id, "disconnect");
+                    }
+                    if (cs->blocked_req && rr->raft && raft_is_leader(rr->raft)) {
```

Comment:
`rr->raft && raft_is_leader(rr->raft)` seems common in these if checks and you can move back to the top if check if you want.